Intermittent interactive test task hang

Had this on macOS today:

```
panic: test timed out after 20m0s
running tests:
	TestInteractive (20m0s)

goroutine 9 [running]:
testing.(*M).startAlarm.func1()
	/Users/task_170723770270584/go1.21.6/go/src/testing/testing.go:2259 +0x320
created by time.goFunc
	/Users/task_170723770270584/go1.21.6/go/src/time/sleep.go:176 +0x38

goroutine 1 [chan receive]:
testing.(*T).Run(0x1400013e1a0, {0x1011c1478?, 0x3bf77d650eaf?}, 0x1012d51a0)
	/Users/task_170723770270584/go1.21.6/go/src/testing/testing.go:1649 +0x350
testing.runTests.func1(0x1400013e1a0?)
	/Users/task_170723770270584/go1.21.6/go/src/testing/testing.go:2054 +0x48
testing.tRunner(0x1400013e1a0, 0x1400005dc28)
	/Users/task_170723770270584/go1.21.6/go/src/testing/testing.go:1595 +0xe8
testing.runTests(0x1400010a960?, {0x1014a77f0, 0x1, 0x1}, {0x40?, 0x1012abc40?, 0x1014b51c0?})
	/Users/task_170723770270584/go1.21.6/go/src/testing/testing.go:2052 +0x3b4
testing.(*M).Run(0x1400010a960)
	/Users/task_170723770270584/go1.21.6/go/src/testing/testing.go:1925 +0x538
main.main()
	_testmain.go:47 +0x1a8

goroutine 6 [IO wait]:
internal/poll.runtime_pollWait(0x1281f66b8, 0x72)
	/Users/task_170723770270584/go1.21.6/go/src/runtime/netpoll.go:343 +0xa0
internal/poll.(*pollDesc).wait(0x14000148100?, 0x14000141000?, 0x0)
	/Users/task_170723770270584/go1.21.6/go/src/internal/poll/fd_poll_runtime.go:84 +0x28
internal/poll.(*pollDesc).waitRead(...)
	/Users/task_170723770270584/go1.21.6/go/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Read(0x14000148100, {0x14000141000, 0x1000, 0x1000})
	/Users/task_170723770270584/go1.21.6/go/src/internal/poll/fd_unix.go:164 +0x200
net.(*netFD).Read(0x14000148100, {0x14000141000?, 0x1282333e0?, 0x100f80230?})
	/Users/task_170723770270584/go1.21.6/go/src/net/fd_posix.go:55 +0x28
net.(*conn).Read(0x1400004c068, {0x14000141000?, 0x14000157be8?, 0x100f6e580?})
	/Users/task_170723770270584/go1.21.6/go/src/net/net.go:179 +0x34
bufio.(*Reader).fill(0x14000074360)
	/Users/task_170723770270584/go1.21.6/go/src/bufio/bufio.go:113 +0xf8
bufio.(*Reader).Peek(0x14000074360, 0x2)
	/Users/task_170723770270584/go1.21.6/go/src/bufio/bufio.go:151 +0x60
github.com/gorilla/websocket.(*Conn).read(0x14000150160, 0x100f64ce4?)
	/Users/task_170723770270584/gopath1.21.6/pkg/mod/github.com/gorilla/websocket@v1.5.1/conn.go:378 +0x28
github.com/gorilla/websocket.(*Conn).advanceFrame(0x14000150160)
	/Users/task_170723770270584/gopath1.21.6/pkg/mod/github.com/gorilla/websocket@v1.5.1/conn.go:824 +0x64
github.com/gorilla/websocket.(*Conn).NextReader(0x14000150160)
	/Users/task_170723770270584/gopath1.21.6/pkg/mod/github.com/gorilla/websocket@v1.5.1/conn.go:1034 +0xfc
github.com/gorilla/websocket.(*Conn).ReadMessage(0x14000206000?)
	/Users/task_170723770270584/gopath1.21.6/pkg/mod/github.com/gorilla/websocket@v1.5.1/conn.go:1120 +0x1c
github.com/taskcluster/taskcluster/v60/workers/generic-worker/interactive.TestInteractive(0x1400013e340)
	/Users/task_170723770270584/taskcluster/workers/generic-worker/interactive/interactive_test.go:52 +0x3c8
testing.tRunner(0x1400013e340, 0x1012d51a0)
	/Users/task_170723770270584/go1.21.6/go/src/testing/testing.go:1595 +0xe8
created by testing.(*T).Run in goroutine 1
	/Users/task_170723770270584/go1.21.6/go/src/testing/testing.go:1648 +0x33c

goroutine 7 [IO wait]:
internal/poll.runtime_pollWait(0x1281f67b0, 0x72)
	/Users/task_170723770270584/go1.21.6/go/src/runtime/netpoll.go:343 +0xa0
internal/poll.(*pollDesc).wait(0x14000148000?, 0x100f6e3a8?, 0x0)
	/Users/task_170723770270584/go1.21.6/go/src/internal/poll/fd_poll_runtime.go:84 +0x28
internal/poll.(*pollDesc).waitRead(...)
	/Users/task_170723770270584/go1.21.6/go/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Accept(0x14000148000)
	/Users/task_170723770270584/go1.21.6/go/src/internal/poll/fd_unix.go:611 +0x250
net.(*netFD).accept(0x14000148000)
	/Users/task_170723770270584/go1.21.6/go/src/net/fd_unix.go:172 +0x28
net.(*TCPListener).accept(0x140000680e0)
	/Users/task_170723770270584/go1.21.6/go/src/net/tcpsock_posix.go:152 +0x28
net.(*TCPListener).Accept(0x140000680e0)
	/Users/task_170723770270584/go1.21.6/go/src/net/tcpsock.go:315 +0x2c
net/http.(*Server).Serve(0x1400014a000, {0x1012d9a90, 0x140000680e0})
	/Users/task_170723770270584/go1.21.6/go/src/net/http/server.go:3056 +0x2b8
net/http/httptest.(*Server).goServe.func1()
	/Users/task_170723770270584/go1.21.6/go/src/net/http/httptest/server.go:310 +0x60
created by net/http/httptest.(*Server).goServe in goroutine 6
	/Users/task_170723770270584/go1.21.6/go/src/net/http/httptest/server.go:308 +0x70

goroutine 18 [select]:
github.com/taskcluster/taskcluster/v60/workers/generic-worker/interactive.(*Interactive).Handler(0x14000026b00, {0x1012d99a0, 0x140000be000}, 0x140000aa000)
	/Users/task_170723770270584/taskcluster/workers/generic-worker/interactive/interactive.go:78 +0x2d0
net/http.HandlerFunc.ServeHTTP(0x1014ea6a0?, {0x1012d99a0?, 0x140000be000?}, 0x14000153b40?)
	/Users/task_170723770270584/go1.21.6/go/src/net/http/server.go:2136 +0x38
net/http.serverHandler.ServeHTTP({0x14000090210?}, {0x1012d99a0?, 0x140000be000?}, 0x6?)
	/Users/task_170723770270584/go1.21.6/go/src/net/http/server.go:2938 +0xbc
net/http.(*conn).serve(0x1400009e000, {0x1012d9e18, 0x140000900f0})
	/Users/task_170723770270584/go1.21.6/go/src/net/http/server.go:2009 +0x518
created by net/http.(*Server).Serve in goroutine 7
	/Users/task_170723770270584/go1.21.6/go/src/net/http/server.go:3086 +0x4cc

goroutine 20 [select]:
os/exec.(*Cmd).watchCtx(0x140000b62c0, 0x140000ba180)
	/Users/task_170723770270584/go1.21.6/go/src/os/exec/exec.go:757 +0x7c
created by os/exec.(*Cmd).Start in goroutine 18
	/Users/task_170723770270584/go1.21.6/go/src/os/exec/exec.go:743 +0x788

goroutine 21 [syscall]:
syscall.syscall6(0x101618a68?, 0x17?, 0x0?, 0x0?, 0x0?, 0x90000101010000?, 0x0?)
	/Users/task_170723770270584/go1.21.6/go/src/runtime/sys_darwin.go:45 +0x68
syscall.wait4(0x140000446b8?, 0x100ff2858?, 0x90?, 0x1012c6120?)
	/Users/task_170723770270584/go1.21.6/go/src/syscall/zsyscall_darwin_arm64.go:43 +0x4c
syscall.Wait4(0x0?, 0x140000446f4, 0x0?, 0x0?)
	/Users/task_170723770270584/go1.21.6/go/src/syscall/syscall_bsd.go:144 +0x28
os.(*Process).wait(0x140000ac0f0)
	/Users/task_170723770270584/go1.21.6/go/src/os/exec_unix.go:43 +0x80
os.(*Process).Wait(...)
	/Users/task_170723770270584/go1.21.6/go/src/os/exec.go:134
os/exec.(*Cmd).Wait(0x140000b62c0)
	/Users/task_170723770270584/go1.21.6/go/src/os/exec/exec.go:890 +0x38
github.com/taskcluster/taskcluster/v60/workers/generic-worker/interactive.CreateInteractiveJob.func1()
	/Users/task_170723770270584/taskcluster/workers/generic-worker/interactive/interactivejob.go:69 +0x28
created by github.com/taskcluster/taskcluster/v60/workers/generic-worker/interactive.CreateInteractiveJob in goroutine 18
	/Users/task_170723770270584/taskcluster/workers/generic-worker/interactive/interactivejob.go:68 +0x30c

goroutine 22 [syscall]:
syscall.syscall(0x1400005ee48?, 0x100fefac0?, 0x800000?, 0x7ffff800000?)
	/Users/task_170723770270584/go1.21.6/go/src/runtime/sys_darwin.go:23 +0x58
syscall.read(0x140000a4180?, {0x14000161000?, 0x0?, 0x2?})
	/Users/task_170723770270584/go1.21.6/go/src/syscall/zsyscall_darwin_arm64.go:1209 +0x48
syscall.Read(...)
	/Users/task_170723770270584/go1.21.6/go/src/syscall/syscall_unix.go:181
internal/poll.ignoringEINTRIO(...)
	/Users/task_170723770270584/go1.21.6/go/src/internal/poll/fd_unix.go:736
internal/poll.(*FD).Read(0x140000a4180, {0x14000161000, 0x1000, 0x1000})
	/Users/task_170723770270584/go1.21.6/go/src/internal/poll/fd_unix.go:160 +0x224
os.(*File).read(...)
	/Users/task_170723770270584/go1.21.6/go/src/os/file_posix.go:29
os.(*File).Read(0x14000092028, {0x14000161000?, 0x14000161000?, 0x0?})
	/Users/task_170723770270584/go1.21.6/go/src/os/file.go:118 +0x70
github.com/taskcluster/taskcluster/v60/workers/generic-worker/interactive.(*InteractiveJob).copyCommandOutputStream(0x140000a8140)
	/Users/task_170723770270584/taskcluster/workers/generic-worker/interactive/interactivejob.go:97 +0x98
created by github.com/taskcluster/taskcluster/v60/workers/generic-worker/interactive.CreateInteractiveJob in goroutine 18
	/Users/task_170723770270584/taskcluster/workers/generic-worker/interactive/interactivejob.go:73 +0x358

goroutine 23 [IO wait]:
internal/poll.runtime_pollWait(0x1281f65c0, 0x72)
	/Users/task_170723770270584/go1.21.6/go/src/runtime/netpoll.go:343 +0xa0
internal/poll.(*pollDesc).wait(0x1400009c000?, 0x140000a6000?, 0x0)
	/Users/task_170723770270584/go1.21.6/go/src/internal/poll/fd_poll_runtime.go:84 +0x28
internal/poll.(*pollDesc).waitRead(...)
	/Users/task_170723770270584/go1.21.6/go/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Read(0x1400009c000, {0x140000a6000, 0x1000, 0x1000})
	/Users/task_170723770270584/go1.21.6/go/src/internal/poll/fd_unix.go:164 +0x200
net.(*netFD).Read(0x1400009c000, {0x140000a6000?, 0x1281b0740?, 0x100f80230?})
	/Users/task_170723770270584/go1.21.6/go/src/net/fd_posix.go:55 +0x28
net.(*conn).Read(0x14000092010, {0x140000a6000?, 0x14000059c18?, 0x0?})
	/Users/task_170723770270584/go1.21.6/go/src/net/net.go:179 +0x34
bufio.(*Reader).fill(0x140000a4000)
	/Users/task_170723770270584/go1.21.6/go/src/bufio/bufio.go:113 +0xf8
bufio.(*Reader).Peek(0x140000a4000, 0x2)
	/Users/task_170723770270584/go1.21.6/go/src/bufio/bufio.go:151 +0x60
github.com/gorilla/websocket.(*Conn).read(0x140000b6160, 0x0?)
	/Users/task_170723770270584/gopath1.21.6/pkg/mod/github.com/gorilla/websocket@v1.5.1/conn.go:378 +0x28
github.com/gorilla/websocket.(*Conn).advanceFrame(0x140000b6160)
	/Users/task_170723770270584/gopath1.21.6/pkg/mod/github.com/gorilla/websocket@v1.5.1/conn.go:824 +0x64
github.com/gorilla/websocket.(*Conn).NextReader(0x140000b6160)
	/Users/task_170723770270584/gopath1.21.6/pkg/mod/github.com/gorilla/websocket@v1.5.1/conn.go:1034 +0xfc
github.com/gorilla/websocket.(*Conn).ReadMessage(0x14000059f48?)
	/Users/task_170723770270584/gopath1.21.6/pkg/mod/github.com/gorilla/websocket@v1.5.1/conn.go:1120 +0x1c
github.com/taskcluster/taskcluster/v60/workers/generic-worker/interactive.(*InteractiveJob).handleWebsocketMessages(0x140000a8140)
	/Users/task_170723770270584/taskcluster/workers/generic-worker/interactive/interactivejob.go:134 +0x9c
created by github.com/taskcluster/taskcluster/v60/workers/generic-worker/interactive.CreateInteractiveJob in goroutine 18
	/Users/task_170723770270584/taskcluster/workers/generic-worker/interactive/interactivejob.go:74 +0x3a4
FAIL	github.com/taskcluster/taskcluster/v60/workers/generic-worker/interactive	1200.027s
ok  	github.com/taskcluster/taskcluster/v60/workers/generic-worker/kc	0.006s
ok  	github.com/taskcluster/taskcluster/v60/workers/generic-worker/livelog	0.116s
ok  	github.com/taskcluster/taskcluster/v60/workers/generic-worker/tcproxy	0.210s
ok  	github.com/taskcluster/taskcluster/v60/workers/generic-worker/win32	0.004s
FAIL
[taskcluster 2024-02-07T11:45:32.826Z]    Exit Code: 1
[taskcluster 2024-02-07T11:45:32.826Z]    User Time: 1m20.047838s
[taskcluster 2024-02-07T11:45:32.826Z]  Kernel Time: 22.205662s
[taskcluster 2024-02-07T11:45:32.826Z]    Wall Time: 22m13.40283s
[taskcluster 2024-02-07T11:45:32.826Z]       Result: FAILED
[taskcluster 2024-02-07T11:45:32.826Z] === Task Finished ===
```

This was for commit 22ce76df4634392906cbc946991b5ed968897106 and seems to be an intermittent issue.
Task log: https://community-tc.services.mozilla.com/tasks/aSU4M3-9TI2MSONCMmpmvQ/runs/0/logs/live/public/logs/live.log

--- a/changelog/issue-6822.md
+++ b/changelog/issue-6822.md
@@ -1,0 +1,3 @@
+level: silent
+reference: issue 6822
+---

--- a/workers/generic-worker/interactive/interactive_test.go
+++ b/workers/generic-worker/interactive/interactive_test.go
@@ -86,7 +86,7 @@ func testInteractive(t *testing.T, port uint16, interactiveCommands InteractiveC
 			t.Fatalf("read error: %v", err)
 		}
 		completeOutput = append(completeOutput, output...)
-		if bytes.Count(completeOutput, expectedBytes) == 3 {
+		if bytes.Count(completeOutput, expectedBytes) == 2 {
 			ok = true
 			break
 		}

--- a/workers/generic-worker/interactive/interactivejob_posix.go
+++ b/workers/generic-worker/interactive/interactivejob_posix.go
@@ -7,8 +7,10 @@ import (
 	"os"
 	"os/exec"
 	"syscall"
+	"time"
 
 	"github.com/creack/pty"
+	"golang.org/x/sys/unix"
 )
 
 type CmdPty struct {
@@ -39,6 +41,29 @@ func (itj *InteractiveJob) Setup(cmd InteractiveCmdType) {
 		close(itj.done)
 	}()
 
+	// Wait for the spawned shell to clear ICANON before returning. Until that
+	// happens, any input we forward from the websocket goes through the
+	// kernel's canonical mode line discipline (line buffering + echo), and the
+	// resulting echo trail depends on which side wins the race between our
+	// write and readline taking over.
+	waitForICANONOff(itj.inner.pty, itj.done)
+}
+
+func waitForICANONOff(p *os.File, done <-chan struct{}) {
+	fd := int(p.Fd())
+	ticker := time.NewTicker(2 * time.Millisecond)
+	defer ticker.Stop()
+	for {
+		t, err := unix.IoctlGetTermios(fd, ioctlGetTermios)
+		if err == nil && t.Lflag&unix.ICANON == 0 {
+			return
+		}
+		select {
+		case <-done:
+			return
+		case <-ticker.C:
+		}
+	}
 }
 
 func (itj *InteractiveJob) resizePty(width uint16, height uint16) error {

--- a/workers/generic-worker/interactive/termios_bsd.go
+++ b/workers/generic-worker/interactive/termios_bsd.go
@@ -1,0 +1,7 @@
+//go:build darwin || freebsd
+
+package interactive
+
+import "golang.org/x/sys/unix"
+
+const ioctlGetTermios = unix.TIOCGETA

--- a/workers/generic-worker/interactive/termios_linux.go
+++ b/workers/generic-worker/interactive/termios_linux.go
@@ -1,0 +1,7 @@
+//go:build linux
+
+package interactive
+
+import "golang.org/x/sys/unix"
+
+const ioctlGetTermios = unix.TCGETS

--- a/workers/generic-worker/interactive_test.go
+++ b/workers/generic-worker/interactive_test.go
@@ -117,7 +117,7 @@ func TestInteractiveCommand(t *testing.T) {
 						t.Fatalf("read error: %v", err)
 					}
 					completeOutput = append(completeOutput, output...)
-					if bytes.Count(completeOutput, expectedBytes) == 3 {
+					if bytes.Count(completeOutput, expectedBytes) == 2 {
 						ok = true
 						break
 					}

